ssh-oqs.c

```diff
@@ -211,7 +280,125 @@ static int ssh_generic_copy_public(const struct sshkey *from, struct sshkey *to)
+{
+  const struct sshkey_impl *impl;
+  int r;
+  impl = oqs_classical_sshkey_impl(k);
+  if ((impl != NULL) && (impl->funcs->generate != NULL)) {
+    if ((r = impl->funcs->generate(k, bits)) != 0) {
+      return r;
+    }
+  }
+  impl = oqs_pq_sshkey_impl(k);
+  if ((r = impl->funcs->generate(k, bits)) != 0) {
+    return r;
+  }
+  return 0;
+}
+
+static int ssh_generic_sign(struct sshkey *key, u_char **sigp,
+        size_t *lenp, const u_char *data, size_t datalen, const char *alg,
+        const char *sk_provider, const char *sk_pin, u_int compat)
+{
+  u_char *sig_classical = NULL, *sig_pq = NULL;
+  size_t len_classical = 0, len_pq = 0;
+  int index = 0;
+  int r;
+  const struct sshkey_impl *impl;
+  if (lenp != NULL) {
+    *lenp = 0;
+  }
+  if (sigp != NULL) {
+    *sigp = NULL;
+  }
+  impl = oqs_pq_sshkey_impl(key);
+  if ((r = impl->funcs->sign(key, &sig_pq, &len_pq, data, datalen, alg,
+                             sk_provider, sk_pin, compat)) != 0) {
+    free(sig_pq);
+    return r;
+  }
+
+  impl = oqs_classical_sshkey_impl(key);
+  if ((impl != NULL) && (impl->funcs->sign != NULL)) {
+    if ((r = impl->funcs->sign(key, &sig_classical, &len_classical, data,
+                               datalen, alg, sk_provider, sk_pin, compat))
+                               != 0) {
+      free(sig_classical);
+      free(sig_pq);
+      return r;
+    }
+    *lenp = 4 + len_classical + 4 + len_pq;
+    if ((*sigp = malloc(*lenp)) == NULL) {
+      free(sig_classical);
+      free(sig_pq);
+      return SSH_ERR_ALLOC_FAIL;
+    }
+    /* encode the classical sig length */
+    POKE_U32(*sigp + index, (size_t) len_classical);
+    index += 4;
+    /* encode the classical sig */
+    memcpy(*sigp + index, sig_classical, (size_t) len_classical);
+    index += len_classical;
+    free(sig_classical);
+    /* encode the PQ sig length */
+    POKE_U32(*sigp + index, len_pq);
+    index += 4;
+    /* encode the PQ sig */
+    memcpy(*sigp + index, sig_pq, len_pq);
+    index += len_pq;
+    free(sig_pq);
+  } else {
+    *sigp = sig_pq;
+    *lenp = len_pq;
+  }
+	return 0;
+}
+
+static int ssh_generic_verify(const struct sshkey *key,
+        const u_char *sig, size_t siglen, const u_char *data, size_t dlen,
+        const char *alg, u_int compat, struct sshkey_sig_details **detailsp)
+{
+  const struct sshkey_impl *impl;
+  const u_char *sig_classical = NULL;
+  size_t siglen_classical = 0;
+  int index = 0;
+  const u_char *sig_pq = NULL;
+  size_t siglen_pq = 0;
+  int r;
+  impl = oqs_classical_sshkey_impl(key);
+  if (impl) {
+    /* classical-PQ hybrid: we separate the signatures */
+    /* decode the classical sig length */
+    siglen_classical = (size_t) PEEK_U32(sig + index);
+    index += 4;
+    /* point to the classical sig */
+    sig_classical = sig + index;
+    index += siglen_classical;
+    /* decode the PQ sig length */
+    siglen_pq = (size_t) PEEK_U32(sig + index);
```

Comment:
Are the lengths here coming from hard-coded values elsewhere or from potentially adversarial input?  If the latter than additional checks would be needed to prevent out-of-bounds reads.